File: sample/cpp/sample_img_hand_gesture.cpp
#include "tdl_model_factory.hpp"

void set_preprocess_parameters(std::shared_ptr<BaseModel> model_hc) {
  PreprocessParams pre_param;

  model_hc->getPreprocessParameters(pre_param);

  pre_param.mean[0] = 2.1179;
  pre_param.mean[1] = 2.0357;
  pre_param.mean[2] = 1.8044;

  pre_param.scale[0] = 0.017126;
  pre_param.scale[1] = 0.017509;
  pre_param.scale[2] = 0.017431;

  model_hc->setPreprocessParameters(pre_param);
}

std::vector<std::shared_ptr<ModelOutputInfo>> extract_crop_hand_landmark(
    std::shared_ptr<BaseModel> model_hc,
    std::vector<std::shared_ptr<BaseImage>> images,
    std::vector<std::shared_ptr<BaseImage>> &hand_crops,
    std::vector<std::shared_ptr<ModelOutputInfo>> &hand_metas) {
  std::shared_ptr<BasePreprocessor> preprocessor = model_hc->getPreprocessor();

  std::vector<std::shared_ptr<ModelOutputInfo>> out_datas;
  for (size_t b = 0; b < images.size(); b++) {
    std::shared_ptr<ModelBoxInfo> hand_meta =
        std::static_pointer_cast<ModelBoxInfo>(hand_metas[b]);

    for (int i = 0; i < hand_meta->bboxes.size(); i++) {
      int x1 = hand_meta->bboxes[i].x1;
      int y1 = hand_meta->bboxes[i].y1;
      int x2 = hand_meta->bboxes[i].x2;
      int y2 = hand_meta->bboxes[i].y2;

      int width = x2 - x1;
      int height = y2 - y1;

      float expansion_factor = 1.125f;
      int new_width = static_cast<int>(width * expansion_factor);
      int new_height = static_cast<int>(height * expansion_factor);

      int crop_x1 = x1 - (new_width - width) / 2;
      int crop_y1 = y1 - (new_height - height) / 2;
      int crop_x2 = crop_x1 + new_width;
      int crop_y2 = crop_y1 + new_height;

      std::shared_ptr<BaseImage> hand_crop = preprocessor->crop(
          images[b], crop_x1, crop_y1, crop_x2 - crop_x1, crop_y2 - crop_y1);

      hand_crops.push_back(hand_crop);
    }
  }
  model_hc->inference(hand_crops, out_datas);

  return out_datas;
}

int main(int argc, char **argv) {
  if (argc != 4) {
    printf("Usage: %s <hand_det_model_path> <hand_classification_model_path>\n",
           argv[0]);
    return -1;
  }
  std::string hd_model_path = argv[1];
  std::string hc_model_path = argv[2];
  std::string image_path = argv[3];

  std::shared_ptr<BaseImage> image1 = ImageFactory::readImage(image_path);
  if (!image1) {
    printf("Failed to create image1\n");
    return -1;
  }

  TDLModelFactory model_factory;

  std::shared_ptr<BaseModel> model_hd =
      model_factory.getModel(ModelType::YOLOV8N_DET_HAND, hd_model_path);
  if (!model_hd) {
    printf("Failed to create model_hd\n");
    return -1;
  }

  std::shared_ptr<BaseModel> model_hc =
      model_factory.getModel(ModelType::CLS_HAND_GESTURE, hc_model_path);
  if (!model_hc) {
    printf("Failed to create model_hc\n");
    return -1;
  }

  // set preprocess parameters
  set_preprocess_parameters(model_hc);

  std::vector<std::shared_ptr<ModelOutputInfo>> out_hd;
  std::vector<std::shared_ptr<BaseImage>> input_images = {image1};
  std::vector<std::shared_ptr<BaseImage>> hand_crops;
  model_hd->inference(input_images, out_hd);
  std::vector<std::shared_ptr<ModelOutputInfo>> out_hc =
      extract_crop_hand_landmark(model_hc, input_images, hand_crops, out_hd);

  char sz_img_name[128];
  for (size_t i = 0; i < out_hc.size(); i++) {
    std::shared_ptr<ModelClassificationInfo> cls_meta =
        std::static_pointer_cast<ModelClassificationInfo>(out_hc[i]);

    printf("hand_crops[%d], label:%d, score:%.2f\n", i,
           cls_meta->topk_class_ids[0], cls_meta->topk_scores[0]);

    sprintf(sz_img_name, "hand_crop_%d_label_%d.jpg", i,
            cls_meta->topk_class_ids[0]);
    ImageFactory::writeImage(sz_img_name, hand_crops[i]);
  }

  return 0;
}

File: sample/python/sample_img_hand_gesture.py
import sys
import os
from tdl import nn, image
import cv2

def crop_and_expand(img, bbox, expansion=1.125):
    x1, y1, x2, y2 = bbox
    width = x2 - x1
    height = y2 - y1
    new_width = int(width * expansion)
    new_height = int(height * expansion)
    crop_x1 = int(x1 - (new_width - width) / 2)
    crop_y1 = int(y1 - (new_height - height) / 2)
    # 边界检查
    crop_x1 = max(0, crop_x1)
    crop_y1 = max(0, crop_y1)
    crop_x2 = min(crop_x1 + new_width, img.shape[1])
    crop_y2 = min(crop_y1 + new_height, img.shape[0])
    w = crop_x2 - crop_x1
    h = crop_y2 - crop_y1
    return img[crop_y1:crop_y1+h, crop_x1:crop_x1+w]

if __name__ == "__main__":
    if len(sys.argv) != 4:
        print("Usage: python3 sample_img_hand_gesture.py <model_path1> <model_path2> <image_path>")
        sys.exit(1)

    model_path1 = sys.argv[1]
    model_path2 = sys.argv[2]
    image_path = sys.argv[3]

    if not os.path.exists(image_path):
        print(f"图像文件不存在: {image_path}")
        sys.exit(1)

    model_config = {
        "mean": (2.1179, 2.0357, 1.8044),
        "scale": (0.017126, 0.017509, 0.017431)
    }

    model_hd = nn.get_model(nn.ModelType.YOLOV8N_DET_HAND, model_path1)
    model_hc = nn.get_model(nn.ModelType.CLS_HAND_GESTURE, model_path2, model_config)

    img = image.read(image_path)
    img_cv = cv2.imread(image_path)

    out_datas = model_hd.inference(img)
    if len(out_datas) == 0:
        print("No object detected")
        sys.exit(0)
    
    hand_crops = []
    for j, obj in enumerate(out_datas):
        bbox = (obj['x1'], obj['y1'], obj['x2'], obj['y2'])
        print(f"hand_bbox_{j}: {bbox}")
        cropped_img = crop_and_expand(img_cv, bbox)
        cropped_img_resized = cv2.resize(cropped_img, (128, 128))
        cv2.imwrite(f"debug_hand_crop_{j}.jpg", cropped_img_resized)
        hand_crops.append(image.Image.from_numpy(cropped_img_resized, image.ImageFormat.BGR_PACKED))
        
    results = []
    for i, hand_img in enumerate(hand_crops):
        out_cls = model_hc.inference(hand_img)
        print(f"hand_crop_{i} out_cls: {out_cls}")
        if out_cls and isinstance(out_cls[0], dict):
            label = out_cls[0].get('topk_class_ids', [out_cls[0].get('class_id', -1)])[0]
            score = out_cls[0].get('topk_scores', [out_cls[0].get('score', 0.0)])[0]
            print(f"hand_crop_{i}, label: {label}, score: {score:.2f}")

            print(f"保存裁剪图像到: hand_crop_{i}_label_{label}.jpg")
        else:
            print(f"hand_crop_{i}, 无法识别手势")

# input: python3 sample_img_hand_gesture.py <model_path1> <model_path2> <image_path>
# output: hand_crop_{i}, label: {label}, score: {score:.2f}